parade/adapters/parsers.py

```
@@ -0,0 +1,240 @@
+            extra = actual_headers - expected_headers
+            msg_parts = []
+            if missing:
+                msg_parts.append(f"missing: {', '.join(missing)}")
+            if extra:
+                msg_parts.append(f"extra: {', '.join(extra)}")
+            msg = f"Invalid CSV headers ({'; '.join(msg_parts)})"
+            raise ParseError(msg)
+
+    def _parse_all_activities(self, reader: csv.DictReader[str]) -> list[UnscheduledActivity]:
+        """Parse all activities from CSV reader.
+
+        Args:
+            reader: CSV reader to parse.
+
+        Returns:
+            List of unscheduled activities.
+
+        Raises:
+            ParseError: If parsing fails.
+            ValidationError: If no activities found.
+        """
+        activities: list[UnscheduledActivity] = []
+        line_num = 2  # Account for header line
+
+        for row in reader:
+            try:
+                activity = self._parse_activity(row)
+                activities.append(activity)
+            except (ParseError, ValidationError):
+                raise
+            except Exception as e:
+                msg = f"Unexpected error: {e}"
+                raise ParseError(msg, line_num) from e
+
+            line_num += 1
+
+        if not activities:
+            msg = "No activities found in CSV"
+            raise ValidationError(msg)
+
+        return activities
+
+    def _create_network(self, activities: list[UnscheduledActivity]) -> UnscheduledProjectNetwork:
+        """Create an unscheduled project network from activities.
+
+        Args:
+            activities: List of activities.
+
+        Returns:
+            An unscheduled project network.
+
+        Raises:
+            ValidationError: If network creation fails validation.
+        """
+        try:
+            return UnscheduledProjectNetwork(activities)
+        except ValueError as e:
+            # Domain-level validation errors
+            raise ValidationError(str(e)) from e
```

This feels bad. Don't wrap this. Just let the exception bubble up.